### ISSUE TITLE
CONFIG_USJ_ENABLE_USB_SERIAL_JTAG can't be disabled (IDFGH-13778)

### ISSUE THREAD

**Issue by higaski** [espressif/esp-idf#14638]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

v5.3

### Operating System used.

Linux

### How did you build your project?

Command line with Make

### If you are using Windows, please specify command line type.

None

### What is the expected behavior?

It should be possible to enable/disable CONFIG_USJ_ENABLE_USB_SERIAL_JTAG.

### What is the actual behavior?

Is it per design that currently the `CONFIG_USJ_ENABLE_USB_SERIAL_JTAG` option (btw, is this a typo or a wordplay on JTAG?) can't be disabled on ESP32-S3 devices?

![menuconfig](https://github.com/user-attachments/assets/b919124a-fe03-4346-873f-eea15b8cf3e7)

### Steps to reproduce.

1. Set up any ESP32-S3 example
2. Try to change the CONFIG_USJ_ENABLE_USB_SERIAL_JTAG option using idf.py menuconfig

**Comment by igrr** [comment]
It can't be disabled because it gets enabled by another option. You can see which one if you press '?', one of the things shown there will be:
```
Symbols currently y-selecting this symbol:
  - ESP_CONSOLE_USB_SERIAL_JTAG_ENABLED
```

and going to that option and pressing '?' again we get
```
Default:
  - y
    Condition (=y):
       ESP_CONSOLE_USB_SERIAL_JTAG(=n)
    || ESP_CONSOLE_SECONDARY_USB_SERIAL_JTAG(=y)
```

So basically, CONFIG_USJ_ENABLE_USB_SERIAL_JTAG gets selected if USB_SERIAL_JTAG is selected as either primary or secondary console output. If you change the console output setting to UART, then you should be able to disable `CONFIG_USJ_ENABLE_USB_SERIAL_JTAG` as well.

**Comment by higaski** [comment]
Thanks!

**Comment by higaski** [comment]
Any updates on this? According to bug reports to me this behavior causes trouble for some Windows 11 users. It seems that switching the device class "on-thy-fly" during initialization somehow confuses Windows 11 and you can end up with no device showing at all.

See https://github.com/hathach/tinyusb/issues/2943

**Comment by igrr** [comment]
@roma-jam since you have looked into the linked issue, could you please check this one as well?

**Comment by higaski** [comment]
Again asking for updates please. Although v5.5 apparently closed some S3 related issues with JTAG and USB I now have a user telling me USB does no longer work for her on Windows 11 at all... she says it gets stuck in JTAG configuration.

**Comment by roma-jam** [comment]
hi @higaski, 

> it seems that switching the device class "on-thy-fly" during initialization somehow confuses Windows 11 and you can end up with no device showing at all.

the USB OTG and USB JTAG share the PHY indeed, but could you please clarify the description: 

>  USB does no longer work for her on Windows 11 at all

Does the user uses the USB OTG functionality? If yes, then which class and Host or Device? If Device, then please share the version of the esp-tinyusb component.

I will check some things and return back ASAP.

**Comment by higaski** [comment]
The following versions are used
```yaml
  idf: "5.5.0"
  espressif/esp_tinyusb:
    version: "1.7.6~1"
    rules:
      - if: "target != linux"
  espressif/tinyusb:
    version: "0.18.0~2"
    rules:
      - if: "target != linux"
```

The code initializes a CDC class with the following kConfig options
```
CONFIG_TINYUSB_TASK_AFFINITY_CPU0=y
CONFIG_TINYUSB_DESC_USE_DEFAULT_PID=n
CONFIG_TINYUSB_DESC_CUSTOM_PID=0x82A5
CONFIG_TINYUSB_DESC_MANUFACTURER_STRING="OpenRemise"
CONFIG_TINYUSB_DESC_PRODUCT_STRING="S3Main"
CONFIG_TINYUSB_DESC_CDC_STRING="ZIMO Interface"
CONFIG_TINYUSB_CDC_ENABLED=y
```

And the snippet doing the initialization
```cpp
  static constexpr tinyusb_config_t tusb_cfg{.device_descriptor = NULL,
                                             .string_descriptor = NULL,
                                             .external_phy = false,
                                             .configuration_descriptor = NULL,
                                             .self_powered = true,
                                             .vbus_monitor_io = vbus_gpio_num};
  ESP_ERROR_CHECK(tinyusb_driver_install(&tusb_cfg));

  static constexpr tinyusb_config_cdcacm_t acm_cfg{
    .usb_dev = TINYUSB_USBDEV_0,
    .cdc_port = TINYUSB_CDC_ACM_0,
    .rx_unread_buf_sz = buffer_size,
    .callback_rx = &tinyusb_cdc_rx_callback,
    .callback_line_state_changed = &tinyusb_cdc_line_state_changed_callback};
  ESP_ERROR_CHECK(tusb_cdc_acm_init(&acm_cfg));
```


Before updating to version 5.5.0 the user told me that USB was working as long as she disabled the flags `self_powered` and `vbus_monitor_io`. Now, she says, it doesn't work at all and her Windows 11 installation only shows the JTAG and never switches to CDC. I know this is all a shot in the dark, unfortunately I can't reproduce the problem on my W11 VM.

**Comment by roma-jam** [comment]
Hi @higaski, 

thanks for the additional info. 

Just in case: are there any options to make the changes in the code and check it with the target hardware, where the problem could be reproduced? 

UPD: What is the board configuration on the user side? Is it devkit or some other hardware? How many USB port connectors are on the board?

**Comment by higaski** [comment]
> Hi [@higaski](https://github.com/higaski),
> 
> thanks for the additional info.
> 
> Just in case: are there any options to make the changes in the code and check it with the target hardware, where the problem could be reproduced?

I'm sorry I'm not sure I understand. If you mean if I could somehow reproduce the issue by changing certain settings then the answer is sadly no. Luckily, the user where it doesn't work is a developer herself, so I'll try to get her involved here.

> UPD: What is the board configuration on the user side? Is it devkit or some other hardware? How many USB port connectors are on the board?

It's a single USB C receptacle, the board is open source so you can see for yourself here: 
https://kicanvas.org/?github=https%3A%2F%2Fgithub.com%2FOpenRemise%2FS3Main

**Comment by roma-jam** [comment]
HI @higaski ,

> I'm sorry I'm not sure I understand.

sorry for not a clear question, but you got it: yes, the idea was to have the possibility to verify the scenario on Win11. So, technically, there is an option to ask the user to add several lines in the esp-idf code, build and flash it and then connect to the Win11. Am I right? 


> It's` a single USB C receptacle, the board is open source so you can see for yourself here:

I have the M5 Stamp module, which has pretty much the same idea where the USB port connector is shared between the USB OTG and USB JTAG. But unfortunately, I am not able to reproduce the problem. But if this related to the same problem as on Mac, the difference in the hardware could also be important.   

As far as I understand, we have two different not related problems so far : 
- Unable to completely disable the USB JTAG
- USB function doesn't changed from the USB JTAG to the USB OTG, when it suppose to do so.

Please, correct me if I am wrong.

**Comment by higaski** [comment]
> sorry for not a clear question, but you got it: yes, the idea was to have the possibility to verify the scenario on Win11. So, technically, there is an option to ask the user to add several lines in the esp-idf code, build and flash it and then connect to the Win11. Am I right?

Yes, absolutely. The person reporting the problem to me even has the toolchain setup and everything. I've already shared the link to this issue here.


> I have the M5 Stamp module, which has pretty much the same idea where the USB port connector is shared between the USB OTG and USB JTAG. But unfortunately, I am not able to reproduce the problem. But if this related to the same problem as on Mac, the difference in the hardware could also be important.
> 
> As far as I understand, we have two different not related problems so far :
> 
>     * Unable to completely disable the USB JTAG
> 
>     * USB function doesn't changed from the USB JTAG to the USB OTG, when it suppose to do so.
> 
> 
> Please, correct me if I am wrong.

Yes I agree. Although I guess solving the first issue would be a valid workaround for the second one. 😄 

I also wonder if either https://github.com/espressif/esp-idf/issues/15912 or https://github.com/espressif/esp-idf/pull/15932 is related to this?

**Comment by roma-jam** [comment]
Hi @higaski , 

> I also wonder if either #15912 or #15932 is related to this?

Highly unlikely, as for this specific case, we need to change the PHY configuration from the USB OTG to JTAG with `usb_phy_new()` call. 

Ok, let's try to eliminate the obvious changes and is that possible to add pull-up override back for the device in here (line 172): 
https://github.com/espressif/esp-idf/blob/b66b5448e0dfac6be135e6f0ac21010aba231180/components/usb/usb_phy.c#L172

So, here is the diff of the changes that need to be added: 

```
diff --git a/components/usb/usb_phy.c b/components/usb/usb_phy.c
index 9432b446cb..41cba55864 100644
--- a/components/usb/usb_phy.c
+++ b/components/usb/usb_phy.c
@@ -170,6 +170,14 @@ esp_err_t usb_phy_otg_set_mode(usb_phy_handle_t handle, usb_otg_mode_t mode)
         esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ONE_INPUT,  otg_sig->bvalid, false);     // HIGH to force USB device mode
         esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ONE_INPUT,  otg_sig->vbusvalid, false);  // receiving a valid Vbus from device
         esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ZERO_INPUT, otg_sig->avalid, false);
+        // Override pull resistors for FS device
+        usb_wrap_pull_override_vals_t vals = {
+            .dp_pu = true,
+            .dm_pu = false,
+            .dp_pd = false,
+            .dm_pd = false,
+        };
+        usb_wrap_hal_phy_enable_pull_override(&handle->wrap_hal, &vals);
     }
```

Please, add the lines, build and flash project. Then try to connect the device to the Win11 Host and share the result. 
Sorry for the inconvenience and thanks for the collaboration.

**Comment by higaski** [comment]
Thanks for pursuing this further. I've already got some feedback.

So apparently with applying the diff USB starts working for her again, but only with the `self_powered` and `vbus_monitor_io` flags disabled.
```cpp
  static constexpr tinyusb_config_t tusb_cfg{.device_descriptor = NULL,
                                             .string_descriptor = NULL,
                                             .external_phy = false,
                                             .configuration_descriptor = NULL,
                                             // .self_powered = true,
                                             // .vbus_monitor_io = vbus_gpio_num
                                            };
```

**Comment by roma-jam** [comment]
Hi @higaski, 

just for info: are you using the light sleep mode in you project? 


Otherwise, there could be a potential requirement to be able to configure these pull resistors based on the hardware specifics.
We will investigate the better way how to make it available and do that asap. 

I'll rename the issue to something which reflects the nature of the problem and keep it open.
Thanks.

**Comment by higaski** [comment]
Not that I know of.

**Comment by tore-espressif** [comment]
@higaski do you have any updates on this?
I reviewed the schematic you shared and everything looks correct. I also reproduced your setup locally, including `.self_powered = true`, and tested on Windows 11 with both USB-A and USB-C host ports — everything worked as expected.

For your hardware, you need to set:
``` c
.vbus_monitor_io = 7
```

**Comment by higaski** [comment]
@tore-espressif 
I'm sorry I don't have any new info. I've only had 20 of those PCBs produced so far so the number of users is small and currently only one reported the behavior described above. 

I forgot to share the definition of `vbus_gpio_num`.
```cpp
constexpr auto vbus_gpio_num{GPIO_NUM_7};
```

So yes, it should be 7.

**Comment by higaski** [comment]
Regardless of the potential Win11 issue, would it be possible to resolve the problem with `CONFIG_USJ_ENABLE_USB_SERIAL_JTAG`?

**Comment by tore-espressif** [comment]
@higaski I belive the original issue has already been resolved/answered. If you are still experiencing a problem, could you please open a new issue with a detailed description? At this point, I’m not sure what the remaining issue is, and a new report would help us track and address it more effectively.

**Comment by higaski** [comment]
@tore-espressif 
I don't think so. `CONFIG_USJ_ENABLE_USB_SERIAL_JTAG` has no effect.

I have it disabled
<img width="562" height="90" alt="Image" src="https://github.com/user-attachments/assets/3449e531-95b7-4822-a189-c15faa193c3f" />

yet my device always initializes JTAG during boot
`Bus 001 Device 013: ID 303a:1001 Espressif USB JTAG/serial debug unit`

**Comment by tore-espressif** [comment]
@higaski I could not reproduce the problem on S3 with default settings. Could you share your full `sdkconfig` file so we could check if there is something else that forces USB-Serial-JTAG on?

For your reference, this section of code disables USB-Serial-JTAG on startup in case it was disabled https://github.com/espressif/esp-idf/blob/2044fba6e71422446986f9ae0909b1ab67e57815/components/esp_system/port/soc/esp32s3/clk.c#L278
you can check if this code is executed in your app eg. by placing warning pragma in your code
``` c
#warning "USB-Serial-JTAG clock and pad disabled"
```

**Comment by higaski** [comment]
@tore-espressif 
The code is executed. So why does my device still show up as `Espressif USB JTAG/serial debug unit`?

Here's my full sdkconfig

[sdkconfig.txt](https://github.com/user-attachments/files/22109943/sdkconfig.txt)

**Comment by tore-espressif** [comment]
If your device shows up as debug unit in your application code, make sure that your app does not call `usb_serial_jtag_ll_enable_bus_clock(true)` or `usb_serial_jtag_ll_phy_enable_pad(true)` anywhere.

> yet my device always initializes JTAG during boot

If it shows up as debug unit only during the 1st and 2nd stage bootloader (but does not once the your application code starts), you can burn
`EFUSE_DIS_USB_SERIAL_JTAG` eFuse to permanently disable USB-Serial-JTAG peripheral. Please note that this step is irreversible

**Comment by higaski** [comment]
Oh I see, so `CONFIG_USJ_ENABLE_USB_SERIAL_JTAG` is essentially just poor naming.

It's more a `CONFIG_USJ_KEEP_USB_SERIAL_JTAG_ENABLED_IN_APP` or something like that. If that's the case, then yes, the issue is kinda resolved.

**Comment by tore-espressif** [comment]
Yes, the USB-Serial-JTAG is enabled by default in the bootloader, so the name is confusing.

In this case we won't rename the symbol (it could break some user's confiurgation) but we can provide better description
